File: hooks/productivity/tdd-guard-check.py
#!/usr/bin/env python3
"""
PreToolUse Hook: TDD Guard Check

Checks if tests exist before allowing writes to production code.
- Warns (doesn't block) if no test file found
- Tracks TDD compliance metrics
- Only checks strategies/ and similar production paths
"""

import json
import os
import sys
from datetime import datetime
from pathlib import Path

# Paths that require tests
PRODUCTION_PATHS = [
    "strategies/",
    "src/",
    "lib/",
    "app/",
    "core/",
    "services/",
    "models/",
    "handlers/",
    "controllers/",
    "utils/",
    "helpers/",
]

# Paths to skip (tests, docs, config)
SKIP_PATHS = [
    "tests/",
    "test_",
    "__pycache__",
    ".pytest_cache",
    "docs/",
    ".claude/",
    "specs/",
    "config/",
    ".planning/",
    "scripts/",
    "migrations/",
    "fixtures/",
]

# Environment variable to control blocking behavior
# Set TDD_GUARD_MODE=warn to only warn (default: block)
BLOCKING_MODE = os.environ.get("TDD_GUARD_MODE", "block") == "block"

# Extensions that need tests
CODE_EXTENSIONS = {".py", ".rs", ".ts", ".js"}

# Metrics storage
METRICS_DIR = Path.home() / ".claude" / "metrics"
TDD_LOG = METRICS_DIR / "tdd_compliance.jsonl"


def get_project_name() -> str:
    """Get project name from git repo or environment."""
    env_name = os.environ.get("CLAUDE_PROJECT_NAME", "")
    if env_name and env_name != "unknown":
        return env_name
    try:
        import subprocess

        result = subprocess.run(
            ["git", "rev-parse", "--show-toplevel"],
            capture_output=True,
            text=True,
            timeout=2,
        )
        if result.returncode == 0:
            return Path(result.stdout.strip()).name
    except Exception:
        pass
    return Path.cwd().name


def log_tdd_metric(data: dict):
    """Log TDD compliance metric."""
    METRICS_DIR.mkdir(parents=True, exist_ok=True)
    entry = {
        "timestamp": datetime.now().isoformat(),
        "project": get_project_name(),
        **data,
    }
    with open(TDD_LOG, "a") as f:
        f.write(json.dumps(entry) + "\n")


def should_check_file(file_path: str) -> bool:
    """Determine if this file needs TDD check."""
    path = Path(file_path)

    # Skip non-code files
    if path.suffix not in CODE_EXTENSIONS:
        return False

    # Skip test files and other non-production paths
    path_str = str(path).lower()
    for skip in SKIP_PATHS:
        if skip in path_str:
            return False

    # Check if in production path
    return any(prod_path in path_str for prod_path in PRODUCTION_PATHS)


def find_test_file(file_path: str) -> Path | None:
    """Find corresponding test file for a source file."""
    path = Path(file_path)
    base_name = path.stem
    parent = path.parent

    # Common test file patterns
    test_patterns = [
        f"test_{base_name}.py",
        f"{base_name}_test.py",
        f"tests/test_{base_name}.py",
        f"tests/{base_name}_test.py",
        f"../tests/test_{base_name}.py",
        f"../../tests/test_{base_name}.py",
        f"../../tests/strategies/test_{base_name}.py",
    ]

    # Check each pattern
    for pattern in test_patterns:
        test_path = parent / pattern
        if test_path.exists():
            return test_path

    # Check in project tests/ directory
    project_dir = Path(os.environ.get("CLAUDE_PROJECT_DIR", "."))
    for tests_dir in ["tests", "test"]:
        test_dir = project_dir / tests_dir
        if test_dir.exists():
            # Search recursively
            for test_file in test_dir.rglob(f"test_{base_name}.py"):
                return test_file
            for test_file in test_dir.rglob(f"{base_name}_test.py"):
                return test_file

    return None


def main():
    try:
        input_data = json.load(sys.stdin)
    except json.JSONDecodeError:
        print(json.dumps({}))
        sys.exit(0)

    tool_name = input_data.get("tool_name", "")
    tool_input = input_data.get("tool_input", {})

    # Only check Write/Edit operations
    if tool_name not in ["Write", "Edit", "MultiEdit"]:
        print(json.dumps({}))
        sys.exit(0)

    file_path = tool_input.get("file_path", "")
    if not file_path:
        print(json.dumps({}))
        sys.exit(0)

    # Check if this file needs TDD verification
    if not should_check_file(file_path):
        log_tdd_metric(
            {
                "type": "skip",
                "file": file_path,
                "reason": "not_production_code",
            }
        )
        print(json.dumps({}))
        sys.exit(0)

    # Look for corresponding test file
    test_file = find_test_file(file_path)

    if test_file:
        # Test exists - good!
        log_tdd_metric(
            {
                "type": "compliant",
                "file": file_path,
                "test_file": str(test_file),
            }
        )
        print(json.dumps({}))
        sys.exit(0)

    # No test found - block or warn based on mode
    log_tdd_metric(
        {
            "type": "violation",
            "file": file_path,
            "test_file": None,
            "blocked": BLOCKING_MODE,
        }
    )

    # Generate message
    base_name = Path(file_path).stem

    if BLOCKING_MODE:
        # BLOCK the operation
        block_message = f"""
🚫 **TDD VIOLATION**: No test file found for `{Path(file_path).name}`

Expected test file patterns:
- `tests/test_{base_name}.py`
- `tests/strategies/test_{base_name}.py`
- `test_{base_name}.py` (same directory)

**Action Required**: Write the test FIRST (Red phase), then implement.

To bypass: Set TDD_GUARD_MODE=warn in environment.
"""
        output = {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "decision": "block",
                "reason": block_message,
            }
        }
        print(json.dumps(output))
        sys.exit(1)  # Non-zero exit to block
    else:
        # Warn only
        warning = f"""
⚠️ **TDD Warning**: No test file found for `{Path(file_path).name}`

Expected test file patterns:
- `tests/test_{base_name}.py`
- `tests/strategies/test_{base_name}.py`
- `test_{base_name}.py` (same directory)

**TDD Best Practice**: Write failing test BEFORE production code.
"""
        output = {
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "notification": warning,
            }
        }
        print(json.dumps(output))
        sys.exit(0)


if __name__ == "__main__":
    main()

File: hooks/productivity/tdd-guard-check.js
#!/usr/bin/env node
/**
 * TDD Guard Hook - Check tests exist before code changes
 *
 * PreToolUse hook that checks if tests exist before allowing writes
 * to production code. Supports strict/warn/off modes.
 *
 * Hook type: PreToolUse (for Write, Edit)
 *
 * Ported from: /media/sam/1TB/claude-hooks-shared/hooks/productivity/tdd-guard-check.py
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const { execSync } = require('child_process');

// Configuration file path
const CONFIG_PATH = path.join(os.homedir(), '.claude', 'tdd-config.json');

// Default configuration
const DEFAULT_CONFIG = {
  mode: 'warn',  // 'strict' | 'warn' | 'off'
  productionPaths: [
    'src/',
    'lib/',
    'app/',
    'core/',
    'services/',
    'models/',
    'handlers/',
    'controllers/',
    'utils/',
    'helpers/',
    'hooks/',
  ],
  skipPaths: [
    'tests/',
    'test/',
    '__tests__/',
    'spec/',
    '__pycache__',
    '.pytest_cache',
    'docs/',
    '.claude/',
    'specs/',
    'config/',
    '.planning/',
    'scripts/',
    'migrations/',
    'fixtures/',
    'node_modules/',
  ],
  codeExtensions: ['.js', '.ts', '.tsx', '.jsx', '.py', '.rs'],
};

// Metrics storage
const METRICS_DIR = path.join(os.homedir(), '.claude', 'metrics');
const TDD_LOG = path.join(METRICS_DIR, 'tdd_compliance.jsonl');

/**
 * Load configuration
 */
function loadConfig() {
  try {
    if (fs.existsSync(CONFIG_PATH)) {
      const content = fs.readFileSync(CONFIG_PATH, 'utf8');
      const config = JSON.parse(content);
      return { ...DEFAULT_CONFIG, ...config };
    }
  } catch {
    // Use defaults
  }
  return DEFAULT_CONFIG;
}

/**
 * Get project name from git repo or environment
 */
function getProjectName() {
  const envName = process.env.CLAUDE_PROJECT_NAME;
  if (envName && envName !== 'unknown') {
    return envName;
  }

  try {
    const result = execSync('git rev-parse --show-toplevel', {
      encoding: 'utf8',
      stdio: ['pipe', 'pipe', 'pipe'],
      timeout: 2000,
    });
    return path.basename(result.trim());
  } catch {
    return path.basename(process.cwd());
  }
}

/**
 * Log TDD compliance metric
 */
function logTddMetric(data) {
  try {
    if (!fs.existsSync(METRICS_DIR)) {
      fs.mkdirSync(METRICS_DIR, { recursive: true });
    }

    const entry = {
      timestamp: new Date().toISOString(),
      project: getProjectName(),
      ...data,
    };

    fs.appendFileSync(TDD_LOG, JSON.stringify(entry) + '\n');
  } catch {
    // Silent fail
  }
}

/**
 * Check if file needs TDD verification
 */
function shouldCheckFile(filePath, config) {
  const ext = path.extname(filePath).toLowerCase();

  // Skip non-code files
  if (!config.codeExtensions.includes(ext)) {
    return false;
  }

  // Normalize path
  const normalizedPath = filePath.replace(/\\/g, '/').toLowerCase();

  // Skip test files and other non-production paths
  for (const skip of config.skipPaths) {
    if (normalizedPath.includes(skip.toLowerCase())) {
      return false;
    }
  }

  // Check if in production path
  return config.productionPaths.some(prodPath =>
    normalizedPath.includes(prodPath.toLowerCase())
  );
}

/**
 * Find corresponding test file for a source file
 */
function findTestFile(filePath) {
  const parsed = path.parse(filePath);
  const baseName = parsed.name;
  const dir = parsed.dir;
  const ext = parsed.ext;

  // Determine test file extension based on source extension
  const testExt = ext || '.js';

  // Common test file patterns
  const testPatterns = [
    // Same directory
    path.join(dir, `${baseName}.test${testExt}`),
    path.join(dir, `${baseName}.spec${testExt}`),
    path.join(dir, `test_${baseName}${testExt}`),
    path.join(dir, `${baseName}_test${testExt}`),
    // __tests__ directory
    path.join(dir, '__tests__', `${baseName}.test${testExt}`),
    path.join(dir, '__tests__', `${baseName}.spec${testExt}`),
    // tests/ directory at same level
    path.join(dir, 'tests', `test_${baseName}${testExt}`),
    path.join(dir, 'tests', `${baseName}.test${testExt}`),
    // Parent tests/ directory
    path.join(dir, '..', 'tests', `test_${baseName}${testExt}`),
    path.join(dir, '..', 'tests', `${baseName}.test${testExt}`),
    // Two levels up tests/ directory
    path.join(dir, '..', '..', 'tests', `test_${baseName}${testExt}`),
    path.join(dir, '..', '..', 'tests', `${baseName}.test${testExt}`),
  ];

  // Check each pattern
  for (const testPath of testPatterns) {
    if (fs.existsSync(testPath)) {
      return testPath;
    }
  }

  // Search in project tests/ directory
  const projectDir = process.env.CLAUDE_PROJECT_DIR || process.cwd();
  const testDirs = ['tests', 'test', '__tests__', 'spec'];

  for (const testDir of testDirs) {
    const testDirPath = path.join(projectDir, testDir);
    if (fs.existsSync(testDirPath)) {
      // Search recursively (one level)
      try {
        const entries = fs.readdirSync(testDirPath, { withFileTypes: true });
        for (const entry of entries) {
          if (entry.isFile()) {
            const name = entry.name.toLowerCase();
            if (name.includes(baseName.toLowerCase()) &&
                (name.includes('test') || name.includes('spec'))) {
              return path.join(testDirPath, entry.name);
            }
          } else if (entry.isDirectory()) {
            // Check subdirectory
            const subDir = path.join(testDirPath, entry.name);
            try {
              const subEntries = fs.readdirSync(subDir);
              for (const subEntry of subEntries) {
                const subName = subEntry.toLowerCase();
                if (subName.includes(baseName.toLowerCase()) &&
                    (subName.includes('test') || subName.includes('spec'))) {
                  return path.join(subDir, subEntry);
                }
              }
            } catch {
              // Ignore
            }
          }
        }
      } catch {
        // Ignore
      }
    }
  }

  return null;
}

/**
 * Read JSON from stdin
 */
async function readStdinJson() {
  return new Promise((resolve, reject) => {
    let data = '';

    process.stdin.setEncoding('utf8');
    process.stdin.on('data', chunk => {
      data += chunk;
    });

    process.stdin.on('end', () => {
      try {
        if (data.trim()) {
          resolve(JSON.parse(data));
        } else {
          resolve({});
        }
      } catch (err) {
        reject(err);
      }
    });

    process.stdin.on('error', reject);

    // Timeout for stdin read
    setTimeout(() => {
      resolve({});
    }, 1000);
  });
}

/**
 * Main function
 */
async function main() {
  try {
    const input = await readStdinJson();
    const config = loadConfig();

    // Check if guard is disabled
    if (config.mode === 'off') {
      console.log(JSON.stringify({}));
      process.exit(0);
    }

    const toolName = input.tool_name || '';
    const toolInput = input.tool_input || {};
    const filePath = toolInput.file_path || '';

    // Only check Write/Edit operations
    if (!['Write', 'Edit', 'MultiEdit'].includes(toolName)) {
      console.log(JSON.stringify({}));
      process.exit(0);
    }

    if (!filePath) {
      console.log(JSON.stringify({}));
      process.exit(0);
    }

    // Check if this file needs TDD verification
    if (!shouldCheckFile(filePath, config)) {
      logTddMetric({
        type: 'skip',
        file: filePath,
        reason: 'not_production_code',
      });
      console.log(JSON.stringify({}));
      process.exit(0);
    }

    // Look for corresponding test file
    const testFile = findTestFile(filePath);

    if (testFile) {
      // Test exists - good!
      logTddMetric({
        type: 'compliant',
        file: filePath,
        test_file: testFile,
      });
      console.log(JSON.stringify({}));
      process.exit(0);
    }

    // No test found - block or warn based on mode
    logTddMetric({
      type: 'violation',
      file: filePath,
      test_file: null,
      blocked: config.mode === 'strict',
    });

    const baseName = path.basename(filePath, path.extname(filePath));
    const ext = path.extname(filePath);
    const testExt = ext || '.js';

    if (config.mode === 'strict') {
      // BLOCK the operation
      const blockMessage = `
TDD VIOLATION: No test file found for \`${path.basename(filePath)}\`

Expected test file patterns:
- \`tests/test_${baseName}${testExt}\`
- \`${baseName}.test${testExt}\` (same directory)
- \`__tests__/${baseName}.test${testExt}\`

Action Required: Write the test FIRST (Red phase), then implement.

To change mode: Edit ~/.claude/tdd-config.json and set "mode": "warn"
`;

      const output = {
        hookSpecificOutput: {
          hookEventName: 'PreToolUse',
          decision: 'block',
          reason: blockMessage,
        },
      };
      console.log(JSON.stringify(output));
      process.exit(1);  // Non-zero exit to block
    } else {
      // Warn only (default)
      const warning = `
TDD Warning: No test file found for \`${path.basename(filePath)}\`

Expected test file patterns:
- \`tests/test_${baseName}${testExt}\`
- \`${baseName}.test${testExt}\` (same directory)
- \`__tests__/${baseName}.test${testExt}\`

TDD Best Practice: Write failing test BEFORE production code.
`;

      const output = {
        hookSpecificOutput: {
          hookEventName: 'PreToolUse',
          decision: 'warn',
          message: warning,
        },
      };
      console.log(JSON.stringify(output));
      process.exit(0);
    }
  } catch (err) {
    // Any error, fail open
    console.log(JSON.stringify({}));
    process.exit(0);
  }
}

main();
